libdrgn/linux_kernel_helpers.c

```diff
@@ -556,380 +536,82 @@ struct drgn_error *linux_helper_find_task(struct drgn_object *res,
 	return err;
 }
 
-struct linux_helper_radix_tree_iter_frame {
-	struct drgn_object slots;
-	uint64_t index;
-	uint64_t shift;
-	uint64_t next_slot;
-};
-
-DEFINE_VECTOR_FUNCTIONS(linux_helper_radix_tree_iter_frame_vector)
-
-struct drgn_error *linux_helper_radix_tree_iter_init(struct linux_helper_radix_tree_iter *iter,
-						     const struct drgn_object *root)
-{
-	struct drgn_program *prog = drgn_object_program(root);
-	iter->started = false;
-	drgn_object_init(&iter->root, prog);
-	drgn_object_init(&iter->entry.node, prog);
-	iter->entry.index = 0;
-	iter->prog = prog;
-
-	struct drgn_error *err =
-		radix_tree_init(prog, root, &iter->RADIX_TREE_INTERNAL_NODE,
-				&iter->RADIX_TREE_MAP_MASK, &iter->node_type, &iter->root);
-
-	if (err) {
-		drgn_object_deinit(&iter->root);
-		drgn_object_deinit(&iter->entry.node);
-		return err;
-	}
-
-	linux_helper_radix_tree_iter_frame_vector_init(&iter->frames);
-	return NULL;
-}
-
-void linux_helper_radix_tree_iter_deinit(struct linux_helper_radix_tree_iter *iter)
-{
-	drgn_object_deinit(&iter->root);
-	drgn_object_deinit(&iter->entry.node);
-	while (iter->frames.size) {
-		drgn_object_deinit(
-			&linux_helper_radix_tree_iter_frame_vector_pop(&iter->frames)->slots);
-	}
-	linux_helper_radix_tree_iter_frame_vector_deinit(&iter->frames);
-}
-
-static struct drgn_error *radix_tree_iter_handle_node(struct linux_helper_radix_tree_iter *iter,
-						      struct drgn_object *_node, uint64_t index,
-						      bool *entry_populated_ret)
+struct drgn_error *
+linux_helper_task_iter_init(struct linux_helper_task_iter *iter,
+			    struct drgn_program *prog)
 {
-	struct drgn_object *node = &iter->entry.node;
 	struct drgn_error *err;
-	uint64_t value;
-
-	err = drgn_object_read(node, _node);
+	iter->done = false;
+	drgn_object_init(&iter->task, prog);
+	err = drgn_program_find_object(prog, "init_task", NULL,
+				       DRGN_FIND_OBJECT_VARIABLE, &iter->task);
 	if (err)
-		return err;
-	err = drgn_object_read_unsigned(node, &value);
-	if (err)
-		return err;
-	if ((value & RADIX_TREE_ENTRY_MASK) != iter->RADIX_TREE_INTERNAL_NODE) {
-		// Base-case, node is NOT internal
-		if (value) {
-			*entry_populated_ret = true;
-			iter->entry.index = index;
-		}
-		return NULL;
-	}
-
-	*entry_populated_ret = false;
-
-	// We are dealing with an internal node, and must iterate over its slots
-
-	err = drgn_object_set_unsigned(node, iter->node_type,
-				       value & ~iter->RADIX_TREE_INTERNAL_NODE, 0);
+		goto out;
+	err = drgn_program_find_type(prog, "struct task_struct", NULL,
+				     &iter->task_struct_type);
```

Comment:
Can you get the type directly from `init_task` rather than looking it up again?

```c
	iter->task_struct_type = drgn_object_qualified_type(&iter->task);
```

Reply:
I think I tried that and got an error, but let me try again